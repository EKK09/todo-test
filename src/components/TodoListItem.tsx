import React, { ChangeEvent } from 'react';
import { useTranslation } from 'react-i18next';
import useTodoListStore, { Todo } from 'src/store/todoList';
import styled from 'styled-components';

export interface TodoListItemProps {
  todo: Todo
}

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  border: 1px solid palevioletred;
  border-radius: 8px;
  margin-bottom: 4px;
  padding: 0 4px;
`;
const Title = styled.h3`
  font-size: 1.5rem;
  color: #727f8b;
`;
const TagsWrapper = styled.div`
  flex-grow: 1;
  display: flex;
  align-items: center;
  justify-content: center;
`;
const Tag = styled.span`
  background-color: CornflowerBlue;
  border-radius: 4px;
  color: white;
  padding: 4px 6px;
  margin: 4px;
`;
const CheckLabel = styled.span`
  color: #727f8b;
  font-weight: bolder;
`;

const TodoListItem = (props: TodoListItemProps): React.ReactElement => {
  const { todo } = props;
  const { t } = useTranslation();
  const updateTodo = useTodoListStore((state) => state.updateTodo);
  const handleCompletedChange = (event: ChangeEvent<HTMLInputElement>) => {
    updateTodo({
      ...todo,
      completed: event.target.checked,
    });
  };

  return (
    <Wrapper>
      <Title>
        {todo.title}
      </Title>
      <TagsWrapper>
        {todo.tags.map((tag) => <Tag key={tag}>{tag}</Tag>)}
      </TagsWrapper>
      <div>
        <CheckLabel>
          {t('todo.completed')}
          <input
            type="checkbox"
            checked={todo.completed}
            onChange={handleCompletedChange}
            data-testid="completed-checkbox"
          />
        </CheckLabel>
      </div>
    </Wrapper>
  );
};

export default TodoListItem;
