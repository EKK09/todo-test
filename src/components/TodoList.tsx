import React, { useEffect } from 'react';
import useTodoListStore from 'src/store/todoList';
import styled from 'styled-components';
import TodoListItem from './TodoListItem';

const LoadingWrapper = styled.div`
  font-weight: bolder;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: nowrap;
`;

const TodoList = (): React.ReactElement => {
  const { todoList, isFetching, fetchTodoList } = useTodoListStore(
    (state) => ({
      todoList: state.todoList,
      isFetching: state.isFetching,
      fetchTodoList: state.fetchTodoList,
    }),
  );
  const todoListItems = todoList.map((todo) => <TodoListItem key={todo.id} todo={todo} />);

  useEffect(() => {
    fetchTodoList();
  }, []);

  if (isFetching) {
    return <LoadingWrapper>Loading ....</LoadingWrapper>;
  }

  return (
    <div>
      {todoListItems}
    </div>
  );
};

export default TodoList;
