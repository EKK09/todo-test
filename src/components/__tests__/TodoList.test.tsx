import React from 'react';
import { render } from '@testing-library/react';
import useTodoListStore, { Todo } from 'src/store/todoList';
import TodoListItem from 'src/components/TodoListItem';
import { create } from 'react-test-renderer';
import TodoList from '../TodoList';

jest.mock('src/components/TodoListItem', () => 'Mock-TodoListItem');

describe('TodoList component', () => {
  const fooTodo: Todo = {
    id: 1,
    title: 'fooTitle',
    tags: ['fooTag'],
    completed: false,
  };
  const barTodo: Todo = {
    id: 2,
    title: 'barTitle',
    tags: ['barTag'],
    completed: true,
  };
  const fooTodoList = [fooTodo, barTodo];
  const mockFetchTodoList = jest.fn();

  beforeEach(() => {
    mockFetchTodoList.mockClear();
  });

  describe('render TodoList', () => {
    it('todo list is not ready', async () => {
      useTodoListStore.setState({ isFetching: true, fetchTodoList: mockFetchTodoList });
      const component = render(<TodoList />);

      expect(component.container).toMatchSnapshot();
    });

    it('todo list is ready', async () => {
      useTodoListStore.setState({ isFetching: false, todoList: fooTodoList });
      const component = render(<TodoList />);

      expect(component.container).toMatchSnapshot();
    });
  });

  it('fetch todo list after component mounted', async () => {
    useTodoListStore.setState({ fetchTodoList: mockFetchTodoList });
    render(<TodoList />);

    expect(mockFetchTodoList).toBeCalledTimes(1);
  });

  it('pass todo to TodoListItem component', async () => {
    useTodoListStore.setState({ fetchTodoList: mockFetchTodoList, todoList: fooTodoList });
    const testRenderer = create(<TodoList />);

    fooTodoList.forEach((todo, index) => {
      const todoListItem = testRenderer.root.findAllByType(TodoListItem)[index];
      expect(todoListItem.props.todo).toEqual(todo);
    });
  });
});
