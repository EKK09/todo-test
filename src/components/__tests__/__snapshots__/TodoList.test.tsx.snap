// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component render TodoList todo list is not ready 1`] = `
<div>
  <div
    class="sc-bdfBQB iunpcb"
  >
    Loading ....
  </div>
</div>
`;

exports[`TodoList component render TodoList todo list is ready 1`] = `
<div>
  <div>
    <mock-todolistitem
      todo="[object Object]"
    />
    <mock-todolistitem
      todo="[object Object]"
    />
  </div>
</div>
`;
